include/random_walks/shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,147 @@
+                    dot_k = -dot_k;
+                }
+
+                int m_fac = static_cast<int>(P.num_of_hyperplanes());
+                Eigen::Matrix<NT, Eigen::Dynamic, 1> Ar(m_fac), Av(m_fac);
+                struct UP { NT inner_vi_ak; int facet_prev; };
+                UP params;
+
+                auto result = P.line_first_positive_intersect(p_, v, Ar, Av, params);
+                NT lambda_hit = result.first;
+                int r = result.second;
+
+                if (!std::isfinite(lambda_hit) || lambda_hit <= NT(0) || lambda_hit > NT(1e3) || r < 0)
+                    continue;
+
+                Point y = p_;
+                for (size_t j = 0; j < dim_; ++j)
+                    y.set_coord(j, p_[j] + lambda_hit * v[j]);
+
+                bool valid = true;
+                for (size_t j = 0; j < dim_; ++j)
+                    if (!std::isfinite(y[j])) { valid = false; break; }
+                if (!valid) continue;
+
+                Eigen::Matrix<NT,1,Eigen::Dynamic> A_row_r = A_.row(r);
+                NT dot_r = A_row_r.dot(v.getCoefficients());
+
+                if (mode_ == Mode::Running) {
+                    p_ = y;
+                    _k = r;
+                    A_row_k_ = A_row_r;
+                    continue;
+                }
+
+                NT beta;
+                if (mode_ == Mode::Original) {
+                    NT den = dot_r - dot_k;
+                    if (std::abs(den) < eps) continue;
+                    beta = dot_r / den;
+                } else { 
+                    beta = -dot_k;
+                }
+
+                if (beta > NT(0) && beta <= NT(1)) {
+                    if (rng.sample_urdist() < beta) {
+                        p_ = y;
+                        _k = r;
+                        A_row_k_ = A_row_r;
+                    }
+                }
+            }
+
+            out_p = p_;
+        }
+
+        const Point& getCurrentPoint() const { return p_; }
+
+    private:
+        template <typename GenericPolytope>
+        void initialize(GenericPolytope const& P, Point const& p0)
+        {
+            dim_        = P.dimension();
+            num_facets_ = P.num_of_hyperplanes();
+            A_ = P.get_mat();
+            b_ = P.get_vec();
+
+            NT eps = std::numeric_limits<NT>::epsilon() * NT(1e4);
+            _k = -1;
+            for (size_t i = 0; i < num_facets_; ++i) {
+                if (std::abs(A_.row(i).dot(p0.getCoefficients()) - b_(i)) < eps) {
+                    _k = int(i);
+                    break;
+                }
+            }
+
+            p_       = p0;
+            A_row_k_ = A_.row(_k);
+        }
+
+        std::size_t dim_{0}, num_facets_{0};
+        Point       p_;
+        int         _k{-1};
+
+        Eigen::Matrix<NT,1,Eigen::Dynamic>         A_row_k_;
+        Eigen::Matrix<NT,Eigen::Dynamic,Eigen::Dynamic> A_;
+        Eigen::Matrix<NT,Eigen::Dynamic,1>              b_;
```

Comment:
Copying b here is ok I'd say. Plz use everywhere `MT` and `VT` by `using MT = ...`